### DISCUSSION TITLE
deinterlacing & double-framerate

### DISCUSSION THREAD

**Discussion by MatejVik** [CESNET/UltraGrid#286] [Q&A]
Hi to all,

I'm testing deinterlacing i uv, `-d gl:d` or `-d gl -p deinterlace` works, but without any visible diference in the output picture. Then i tryed `-d gl: -p double-framerate` and `-d gl:d-p double-framerate`  both not working with message `Unknown postprocess module: double-framerate`. I tryed master a continuous builds.

questions:
1. Is the problem on my chair, or olsewhere? 
2. Is there some way to deinterace for nice looking output, something like YDIF?

**interlaced ndi camera side**

./uv -f A:mult:3 -f V:rs:200:250 -t ndi:name="birddog12 (CAM)":url=172.19.40.119:5962:bandwidth=100:color=100 -c libavcodec:encoder=hevc_nvenc:preset=slow:profile=main:subsampling=420:bitrate=20M --param lavc-rc-buffer-size-factor=0 -d dummy --param force-lavd-decoder=hevc_cuvid -s portaudio:59 --audio-codec opus:bitrate=64K -m 9000 -P 8000:7000:8001:7001 192.168.1.30
UltraGrid 1.8+ (master rev 910a522 built Jan  3 2023 15:59:29)

Display device   : dummy
Capture device   : ndi
Audio capture    : portaudio:59
Audio playback   : none
MTU              : 9000 B
Video compression: libavcodec:encoder=hevc_nvenc:preset=slow:profile=main:subsampling=420:bitrate=20M
Audio codec      : OPUS
Network protocol : UltraGrid RTP
Audio FEC        : mult:3
Video FEC        : rs:200:250

[NAT] Private outbound IPv4 address detected and binding as a receiver. Consider adding '-N' option for NAT traversal.
[PortAudio] Using PortAudio V19.7.0-devel, revision 147dd722548358763a8b649b3e4b41dfffbcfbb6
Initializing portaudio capture.
Using PortAudio version: PortAudio V19.7.0-devel, revision 147dd722548358763a8b649b3e4b41dfffbcfbb6
Using input audio device:DVS Receive  1-2 (Dante Virtual Soundcard) (output channels: 0; input channels: 2; Windows WASAPI)
RTP: rx_port should be even, using 8001.
RTP: tx_port should be even, using 7001.
[lavcd aud.] Using audio encoder: libopus
This application uses NDI® available from http://ndi.tv/
NDI® is a registered trademark of NewTek, Inc.

172.19.40.119:5962Control socket listening on port 63891
[control] Fec changed successfully
Audio sending started.
[lavcd aud.] Using audio encoder: libopus
[NDI cap.] Looking for source(s)...
    Last message repeated 7 times
[NDI cap.] Receiving from source: EC-4-P200 (CAM), URL: 172.19.40.119:5962
[NDI cap.] Status changed!
    Last message repeated 1 times
[NDI cap.] Received video changed: 1920x1080 @60.00i, codec UYVY
[lavc] Using codec: H.265, encoder: hevc_nvenc
[lavc] Setting bitrate to 20.0 Mbps.
[lavc] FFmpeg not patched, not enabling Intra Refresh.
[lavc] Selected pixfmt: yuv420p
FEC symbol size: 1006, symbols per packet: 8, payload size: 8048
FEC symbol size: 321, symbols per packet: 27, payload size: 8667
FEC symbol size: 381, symbols per packet: 23, payload size: 8763
[Audio sender] Sent 229440 samples in last 5.001879 seconds.
[Audio sender] Volume: [0] -39.17/-25.73, [1] -39.17/-25.73 dBFS RMS/peak
[NDI cap.] 140 frames in 5.01715 seconds = 27.9043 FPS

**display deinterlacing side**
PS C:\UltraGrid-continuous-win64> ./uv -f A:mult:4 -t screen:fps=50 -c libavcodec:encoder=hevc_qsv:preset=slow:profile=main:bitrate=15M -d gl:single:d -p double-framerate --param force-lavd-decoder=hevc_qsv -r wasapi --param audio-buffer-len=72000 -P 7000:8000:7001:8001 192.168.1.20
UltraGrid 1.8+ (master rev 910a522 built Jan  3 2023 15:59:29)

Display device   : gl
Capture device   : screen
Audio capture    : none
Audio playback   : wasapi
MTU              : 1500 B
Video compression: libavcodec:encoder=hevc_qsv:preset=slow:profile=main:bitrate=15M
Audio codec      : PCM
Network protocol : UltraGrid RTP
Audio FEC        : mult:4
Video FEC        : none

[NAT] Private outbound IPv4 address detected and binding as a receiver. Consider adding '-N' option for NAT traversal.
[WASAPI play.] Using device: Speakers / Headphones (Realtek Audio)
RTP: rx_port should be even, using 7001.
RTP: tx_port should be even, using 8001.
[GL] Single-buffering is not recommended and may not work, also GLFW discourages its use.
GL setup: fullscreen: OFF, deinterlace: ON
[GL] OpenGL 2.0 is supported...
[GL] Buffer depth - R: 8b, G: 8b, B: 8b
Unknown postprocess module: double-framerate
Exit
PS C:\UltraGrid-continuous-win64>

**Comment by MartinPulec** [comment]
Thanks very much for reporting. The thing is that this stuff is a bit old and not so often used and it needs some house-keeping. I had to do some fixes as well, which there are already present in the continuous builds and I'll add something more tomorrow. My remarks.

1. `double-framerate` in wiki was incorrect, it should be actually `double_framerate` (underscore instead of hyphen). It also needed some small fixes.
2. deinterlacing works for some codecs only (RGBA, RGB and UYVY). What is worse, if it doesn't, it is silently not performed. I've added a check and error message that will notify the user. I'll add support for many more of the pixel formats tomorrow.
3. actually, from the point 2 I am not sure why it didn't deinterlace for you, if you used _UYVY_
4. yadif or other more sophisticated algorithms are possible, maybe it would be even sufficient ordinary linear; `-p double-framerate:d` works slightly differently. But it produced acceptable results for us back then.

**Reply by MartinPulec** [reply]
As promised, there arrived some small updates on that in the development version - deinterlace (blend) now supports almost (if not) all pixel formats, so it should always work. I've also added some other deinterlacing modes (classic linear + bob) and updated the particular [wiki page](https://github.com/CESNET/UltraGrid/wiki/Interlaced-Input).

update 13th Jan: there are another improvements (mainly) on blend deinterlace - improved performance and quality (of course, it is still blend)

**Reply by MatejVik** [reply]
Thanks so much, you're seriously quick. I'll try it asap, now I'm on other types of events without test equipment.
